### ISSUE TITLE
Unable to remove member from cluster if member not running

### ISSUE THREAD

**Issue by bcwaldon** [coreos/etcd#1483]
1. Start a single-node cluster

```
./bin/etcd -name=uno -initial-cluster=uno=http://127.0.0.1:2380
```

2. Add a member to the cluster

```
% curl -X POST http://127.0.0.1:4001/v2/admin/members -H "Content-Type: application/json" -d '{"peerURLs":["http://127.0.0.1:8080"]}'
{"id":9178644553537991936,"peerURLs":["http://127.0.0.1:8080"]}
```

3. Note the log on the first node

```
2014/10/29 13:53:39 etcdhttp: added node 7f6118974534b500 with peer urls [http://127.0.0.1:8080]
```

4. Try to remove the node

```
% curl -i -X DELETE http://127.0.0.1:4001/v2/admin/members/7f6118974534b500
HTTP/1.1 500 Internal Server Error
Content-Type: text/plain; charset=utf-8
X-Etcd-Cluster-Id: 9dbd6201abd472d6
Date: Wed, 29 Oct 2014 20:54:42 GMT
Transfer-Encoding: chunked

Internal Server Error
```

5. Note the logs on the first node

```
2014/10/29 13:54:42 etcdhttp: remove node 7f6118974534b500
2014/10/29 13:54:42 etcdhttp: error removing node 7f6118974534b500: etcdserver: request timed out
2014/10/29 13:54:42 etcdserver: request timed out
```

**Comment by jakedt** [comment]
I am seeing the exact same behavior with a 3 node cluster with 2 healthy nodes. If I attempt to remove a node for a machine which is no longer running, and therefore listening on the ip address, I will get a request timed out and an Internal Server Error. Here are some logs:

```json
{
  "members": [
    {
      "id": "19dd924a6bb20a14",
      "name": "i-c61b106a",
      "peerURLs": [
        "http://10.0.2.216:2380"
      ],
      "clientURLs": [
        "https://52.20.15.247:2379"
      ]
    },
    {
      "id": "2eb577a8ab57f628",
      "name": "i-382864ea",
      "peerURLs": [
        "http://10.0.1.65:2380"
      ],
      "clientURLs": [
        "https://52.6.203.172:2379"
      ]
    },
    {
      "id": "9b20587f244fdf6f",
      "name": "",
      "peerURLs": [
        "http://10.0.3.123:2380"
      ],
      "clientURLs": []
    }
  ]
}
```

```sh
curl -k --cert etcd_client.crt --key etcd_client.key "https://10.0.1.65:2379/v2/members/9b20587f244fdf6f" -XDELETE
```

```
Aug 20 15:43:24 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:24 etcdhttp: error removing node 9b20587f244fdf6f: etcdserver: request timed out
Aug 20 15:43:24 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:24 etcdhttp: unexpected error: etcdserver: request timed out
```

Our hypothesis is that etcd is attempting to contact the bad peer and ask it to remove itself from the cluster, but when it cannot be reached (no machine is listening) the request times out. This is unanticipated, so the cluster reports the error to the client. When checking the cluster membership the node is actually removed though, despite the 500 error.

```json
{
  "members": [
    {
      "id": "19dd924a6bb20a14",
      "name": "i-c61b106a",
      "peerURLs": [
        "http://10.0.2.216:2380"
      ],
      "clientURLs": [
        "https://52.20.15.247:2379"
      ]
    },
    {
      "id": "2eb577a8ab57f628",
      "name": "i-382864ea",
      "peerURLs": [
        "http://10.0.1.65:2380"
      ],
      "clientURLs": [
        "https://52.6.203.172:2379"
      ]
    }
  ]
}
```

**Comment by jakedt** [comment]
More logs from after the event showing leader elections getting queued up, and the node eventually being removed.
```
Aug 20 15:43:27 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:27 raft: 2eb577a8ab57f628 is starting a new election at term 138
Aug 20 15:43:27 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:27 raft: 2eb577a8ab57f628 became candidate at term 139
Aug 20 15:43:27 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:27 raft: 2eb577a8ab57f628 received vote from 2eb577a8ab57f628 at term 139
Aug 20 15:43:27 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:27 raft: 2eb577a8ab57f628 [logterm: 133, index: 11333] sent vote request to 19dd924a6bb20a14 at term 139
Aug 20 15:43:27 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:27 raft: 2eb577a8ab57f628 [logterm: 133, index: 11333] sent vote request to 9b20587f244fdf6f at term 139
Aug 20 15:43:29 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:29 raft: 2eb577a8ab57f628 is starting a new election at term 139
Aug 20 15:43:29 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:29 raft: 2eb577a8ab57f628 became candidate at term 140
Aug 20 15:43:29 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:29 raft: 2eb577a8ab57f628 received vote from 2eb577a8ab57f628 at term 140
Aug 20 15:43:29 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:29 raft: 2eb577a8ab57f628 [logterm: 133, index: 11333] sent vote request to 19dd924a6bb20a14 at term 140
Aug 20 15:43:29 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:29 raft: 2eb577a8ab57f628 [logterm: 133, index: 11333] sent vote request to 9b20587f244fdf6f at term 140
Aug 20 15:43:31 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:31 raft: 2eb577a8ab57f628 is starting a new election at term 140
Aug 20 15:43:31 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:31 raft: 2eb577a8ab57f628 became candidate at term 141
Aug 20 15:43:31 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:31 raft: 2eb577a8ab57f628 received vote from 2eb577a8ab57f628 at term 141
Aug 20 15:43:31 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:31 raft: 2eb577a8ab57f628 [logterm: 133, index: 11333] sent vote request to 19dd924a6bb20a14 at term 141
Aug 20 15:43:31 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:31 raft: 2eb577a8ab57f628 [logterm: 133, index: 11333] sent vote request to 9b20587f244fdf6f at term 141
Aug 20 15:43:32 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:32 raft: 2eb577a8ab57f628 is starting a new election at term 141
Aug 20 15:43:32 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:32 raft: 2eb577a8ab57f628 became candidate at term 142
Aug 20 15:43:32 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:32 raft: 2eb577a8ab57f628 received vote from 2eb577a8ab57f628 at term 142
Aug 20 15:43:32 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:32 raft: 2eb577a8ab57f628 [logterm: 133, index: 11333] sent vote request to 19dd924a6bb20a14 at term 142
Aug 20 15:43:32 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:32 raft: 2eb577a8ab57f628 [logterm: 133, index: 11333] sent vote request to 9b20587f244fdf6f at term 142
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 is starting a new election at term 142
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 became candidate at term 143
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 received vote from 2eb577a8ab57f628 at term 143
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 [logterm: 133, index: 11333] sent vote request to 19dd924a6bb20a14 at term 143
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 [logterm: 133, index: 11333] sent vote request to 9b20587f244fdf6f at term 143
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 [term: 143] ignored a MsgVoteResp message with lower term from 19dd924a6bb20a14 [term: 134]
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 [term: 143] ignored a MsgVoteResp message with lower term from 19dd924a6bb20a14 [term: 137]
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 [term: 143] ignored a MsgVoteResp message with lower term from 19dd924a6bb20a14 [term: 136]
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 [term: 143] ignored a MsgVoteResp message with lower term from 19dd924a6bb20a14 [term: 135]
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 [term: 143] ignored a MsgVoteResp message with lower term from 19dd924a6bb20a14 [term: 138]
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 [term: 143] ignored a MsgVoteResp message with lower term from 19dd924a6bb20a14 [term: 139]
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 [term: 143] ignored a MsgVoteResp message with lower term from 19dd924a6bb20a14 [term: 140]
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 [term: 143] ignored a MsgVoteResp message with lower term from 19dd924a6bb20a14 [term: 141]
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 [term: 143] ignored a MsgVoteResp message with lower term from 19dd924a6bb20a14 [term: 142]
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 received vote from 19dd924a6bb20a14 at term 143
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 [q:2] has received 2 votes and 0 vote rejections
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft: 2eb577a8ab57f628 became leader at term 143
Aug 20 15:43:34 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:34 raft.node: 2eb577a8ab57f628 elected leader 2eb577a8ab57f628 at term 143
Aug 20 15:43:36 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:36 raft: 2eb577a8ab57f628 [term: 143] received a MsgVote message with higher term from 19dd924a6bb20a14 [term: 144]
Aug 20 15:43:36 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:36 raft: 2eb577a8ab57f628 became follower at term 144
Aug 20 15:43:36 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:36 raft: 2eb577a8ab57f628 [logterm: 143, index: 11338, vote: 0] voted for 19dd924a6bb20a14 [logterm: 143, index: 11338] at term 144
Aug 20 15:43:36 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:36 raft.node: 2eb577a8ab57f628 lost leader 2eb577a8ab57f628 at term 144
Aug 20 15:43:37 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:37 etcdserver: removed member 9b20587f244fdf6f from cluster 3798e4719398a84b
Aug 20 15:43:37 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:37 rafthttp: server streaming to 19dd924a6bb20a14 at term 143 has been stopped
Aug 20 15:43:37 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:37 raft.node: 2eb577a8ab57f628 elected leader 19dd924a6bb20a14 at term 144
Aug 20 15:43:37 ip-10-0-1-65.ec2.internal etcd2[861]: 2015/08/20 15:43:37 rafthttp: starting client stream to 19dd924a6bb20a14 at term 144
```

**Comment by xiang90** [comment]
@jakedt What version of etcd are you running? This seems to be normal. You were experiencing leader election. The removal request did timed out.

**Comment by jakedt** [comment]
Getting back an Internal Server Error with a 500 when removing a bad peer does not seem normal. Having that peer successfully removed after receiving a 500, also does not seem normal.

**Comment by xiang90** [comment]
@jakedt We have improved the logging around unexpected timeout: see https://github.com/coreos/etcd/pull/3277

**Comment by xiang90** [comment]
@jakedt Your server were experiencing leader election for quite a long time for some reason.

etcd cannot know if the request will be committed or not. It timed out. 

I think this is a reasonable behavior.

**Comment by xiang90** [comment]
@jakedt After offline discussion, we found out the root cause. There is a bug in etcd 2.0.11 that can lead to this behavior. See https://github.com/coreos/etcd/issues/2888. It has been fixed since 2.1 release.
